workflows/cloudify_system_workflows/dsl_import_resolver/resolver_with_catalog_support.py

```diff
@@ -130,6 +141,83 @@ def _resolve_plugin_yaml_url(self, import_url):
         plugin = self._find_plugin(name, plugin_filters)
         return self._make_plugin_yaml_url(plugin)
 
+    @staticmethod
+    def _create_zip(source, destination, include_folder=True):
+        with closing(zipfile.ZipFile(destination, 'w')) as zip_file:
+            for root, _, files in os.walk(source):
+                for filename in files:
+                    file_path = os.path.join(root, filename)
+                    source_dir = os.path.dirname(source) if include_folder \
+                        else source
+                    zip_file.write(
+                        file_path, os.path.relpath(file_path, source_dir))
+        return destination
+
+    @staticmethod
+    def _download_file(url, dest_path, target_filename=None):
+        with requests.get(url, stream=True, timeout=(5, None)) as resp:
+            resp.raise_for_status()
+            if not target_filename:
+                target_filename = os.path.basename(url)
+            file_path = os.path.join(dest_path, target_filename)
+            with open(file_path, 'wb') as f:
+                for chunk in resp.iter_content(chunk_size=8192):
+                    if chunk:
+                        f.write(chunk)
+        return file_path
+
+    def _upload_missing_plugin(self, name, specifier_set, distribution):
+        plugins_data = urllib.request.urlopen(PLUGIN_CATALOG_URL).read()
+        plugins_data = json.loads(plugins_data)
+        requested_plugin_data = [x for x in plugins_data if x["name"] == name]
+
+        matching_versions = []
+        for plugin in requested_plugin_data:
+            p_icon = plugin['icon']
+            matching_versions.extend([
+                (v, parse_version(v)) for v in plugin['versions'].keys()
+                if (parse_version(v) in specifier_set
+                    and plugin['versions'][v]['yaml']
+                    and plugin['versions'][v]['wagons'])
+            ])
+        if not matching_versions:
+            raise FileNotFoundError
+
+        max_item = max(matching_versions, key=lambda v_p: v_p[1])
+        matching_plugin_data = plugin['versions'][max_item[0]]
+
+        p_yaml = matching_plugin_data['yaml']
+        p_wagons = [x['url'] for x in matching_plugin_data['wagons'] if
+                    x['name'].lower() == distribution]
+
+        if not p_wagons:
+            raise FileNotFoundError
+        p_wagon = p_wagons[0]
+
+        plugin_target_path = tempfile.mkdtemp()
+        self._download_file(p_yaml, plugin_target_path)
+        self._download_file(p_wagon, plugin_target_path)
+        if p_icon:
+            self._download_file(p_icon, plugin_target_path, 'icon.png')
+
+        plugin_zip = plugin_target_path + '.zip'
+        self._create_zip(plugin_target_path, plugin_zip,
+                         include_folder=False)
+        shutil.rmtree(plugin_target_path)
+        self.client.plugins.upload(plugin_zip)
```

Comment:
Shouldn't we delete the plugin_zip after this?

Reply:
yeah we can I guess